### DISCUSSION TITLE
Tutorial for nested ww3 runs for ww3 6.07

### DISCUSSION THREAD

**Discussion by JMSnyder1980** [NOAA-EMC/WW3#959] [General]
Are there any tutorials or examples on how to do a full nested run off ww3 first starting with a broad nest that has boundary conditions for a smaller nest?  I cannot find any such thing in the regtests and would like an example of how to do this so I can apply to a global nest going into a smaller regional nest.

**Comment by EmilioEchevarria** [comment]
There's a couple of ways to accomplish this, and the manual discusses them, see here: https://github.com/NOAA-EMC/WW3/blob/develop/manual/app/nest.tex

One option is to run your global grid specifying that you want boundary conditions for a regional (nested) grid. You define that in ww3_grid.inp (see [here](https://github.com/NOAA-EMC/WW3/blob/develop/model/inp/ww3_grid.inp#L832-L848) for example). Also make sure you output boundary data (see [here](https://github.com/NOAA-EMC/WW3/blob/develop/model/inp/ww3_shel.inp#L323-L327)). Once this is set up, run your global grid and it should generate a file `nest1.ww3`, rename that to `nest.ww3` and put it in the directory where you're going to run your regional grid.

Another option, use spectral output data (it could be from your global simulation) and use [ww3_bounc](https://github.com/NOAA-EMC/WW3/blob/develop/model/inp/ww3_bounc.inp) to generate the file `nest.ww3` to be used in your regional run.

Hope that helps. 
Emilio

**Reply by JMSnyder1980** [reply]
I am having problems applying the ww3_grid.inp to a global bathemetry file.   How do I modify the ww3_grid.inp to be able to be compatible with the global bathemetry files?  Can you guide me in thiis or help modify the file to work successfully with the bathemetry files I am sending you?
[global_1p0_mask_new.dat.txt](https://github.com/NOAA-EMC/WW3/files/11074891/global_1p0_mask_new.dat.txt)
[global_1p0_obst_new.dat.txt](https://github.com/NOAA-EMC/WW3/files/11074893/global_1p0_obst_new.dat.txt)
[global_1p0_bathy_new.dat.txt](https://github.com/NOAA-EMC/WW3/files/11074895/global_1p0_bathy_new.dat.txt)

**Reply by EmilioEchevarria** [reply]
I will try to have a look later and run ww3_grid with these files. But please tell me, which error are you getting?

**Reply by JMSnyder1980** [reply]
This is the error I am getting:


  Output boundary points :
 --------------------------------------------------

 *** WAVEWATCH III ERROR IN W3GRID :
     ERROR IN READING FROM INPUT FILE
     IOSTAT =    0

**Reply by EmilioEchevarria** [reply]
Ok I just saw your other question, the `ww3_grid.inp` you're using seems to be taken from [here](https://github.com/NOAA-EMC/WW3/blob/develop/model/inp/ww3_grid.inp), which is a test. Naturally this won't work in a different setting with a different grid. The idea is that you define `ww3_grid.inp` based on your specific grid and specific configuration you want to use (for example, if you are using regular grids, unstructured grids, SMC grids, etc. there are options for each of these in `ww3_grid.inp` and you need to define them based on your case). [This](https://github.com/NOAA-EMC/WW3/blob/develop/model/inp/ww3_grid.inp) can be taken as a set of instructions that you can follow to generate your own input file correctly. 

The error you got indicates that there is an issue with your `ww3_grid.inp` file and WW3 failed trying to read it. If you keep getting it, it means you need to change your input file to make it have the correct format that WW3 expects. 

I haven't tested your files, but I imagine you would need something like this maybe:
```
$ -------------------------------------------------------------------- $
  'Name of your grid'
$
  1.100000 0.041180 25 24 0.
$
  F T T T F T 
$
  1800  600  900  30
$
   &MISC FLAGTR = 4, CICE0 = 0.25, CICEN = 0.75/
   END OF NAMELISTS
  'RECT' T 'NONE'
  361	 159 
  1.0	 1.0  1.00 
  0   -78   1.00
  -0.10   1.0  20  1.0  1  1 '(....)'  'NAME' '/path/to/your/depth/file' 
  30  1.0  1  1  '(....)'  'NAME'  '/path/to/your/obst/file'
  40            1  1  '(....)'  'NAME'  '/path/to/your/mask/file'
$
$ Nest boundary conditions
$
nest_xW nest_yN 0.0 -DY -NY
nest_xW nest_yS DX 0.0 NX
nest_xE nest_yS 0.0 DY NY
nest_xE nest_yN -DX 0.0 NX
$
$
      0.      0.    0.    0.       0
$
$ ----------------- 
$ End of input file
$ ----------------- 
```

But again, you need to work it out for your own grid and configuration.

**Reply by CarstenHansen** [reply]
The namelists version of the ww3_grid input, [ww3_grid.nml](https://github.com/NOAA-EMC/WW3/blob/develop/model/nml/ww3_grid.nml), contains more lately updated instructions. It may for example be instructive to read the section here about how to 'Define the timesteps parameterization'. I don't remember if this is reflected in the manual.

Judging from your global bathymetry data, it seems that you need to set a scaling (division) factor of value 0.1 rather than the value 1.0 in @EmilioEchevarria's example.

**Comment by abayhaqi** [comment]
Hi @EmilioEchevarria I have successfully produced the two nest files (nest1.ww3 and nest2.ww3) following your answer. However, I am just wondering how to rename it to a nest.ww3. Then It could be two files with the same name  (?)

**Reply by EmilioEchevarria** [reply]
Hi @abayhaqi , can you please share your ww3_grid.inp file? nest1.ww3 and nest2.ww3 means you defined nesting output for 2 "child" grids. Now when you are going to run WW3 for one of these grids (say nest1) you need to run it in a different directory, and copy nest1.ww3 to that directory (simply cp nest1.ww3 /path/to/nest_run/nest.ww3)

And when you are going to run nest2 simulation, again you do that in a different directory and copy nest2.ww3 to that directory as nest.ww3

**Reply by abayhaqi** [reply]
Here, I attach my ww3_grid.inp. I want to make boundaries in south, east and north. Please correct me if I put a wrong way in ww3_grid. I am new in WW3.

[ww3_grid.txt](https://github.com/NOAA-EMC/WW3/files/12425374/ww3_grid.txt)

**Reply by EmilioEchevarria** [reply]
Regarding the nesting issue, (if I interpret what you're trying to do correctly) I would modify lines 645-647 in your ww3_grid.inp to something like:
```
110 54.5 0 -1 -34
110 20.5 1 0 41
151 20.5 0 1 34
151 54.5 -1 0 41
```
This should give you this nest grid:
![image](https://github.com/NOAA-EMC/WW3/assets/36142422/71a12518-1267-42d4-afa3-048a9629ebcb)

The explanation for this is [here](https://github.com/NOAA-EMC/WW3/blob/25bc4d7a21752587dd6f964ab49a0a2722acbcba/model/inp/ww3_grid.inp#L844-L846). In the example above, we generate 4 segments at which we'll extract wave boundary conditions, these 4 segments correspond to the edges of your child grid. First line: We start at point X=110 Y=54.5, and we have a DX=0 and DY=-1, with 34 points along that line (the negative 34 means we start a new nest#.ww3 file with this). With DX=0 and DY=-1 (southward), that takes us to the point X=110 Y=20.5. So in the second line: we start at point X=110 Y=20.5, and we move with DX=1 and DY=0 (eastward), with 41 points along this segment, that takes us to point X=151 Y=20.5. Similarly in the third line we move northward, in the forth line we move westward, closing the grid. 

Hope this helps

**Reply by abayhaqi** [reply]
Ahh, I see. Thank you so much. It helps me a lot

**Reply by abayhaqi** [reply]
Hi @EmilioEchevarria , 
I am just wondering, when I run the regional model, is there something that I should define in ww3_shel. inp or ww3_grid.inp to recognize the nest.ww3 file from the global run? or will it be automatically recognized when I copy the nest.ww3 into the regional model directory? 
Thank you

**Reply by EmilioEchevarria** [reply]
Hi @abayhaqi, you should also modify the mask of your child grid to have 2s at the edges of the grid (or, more generally, wherever you want wave boundary conditions). See [here](https://github.com/NOAA-EMC/WW3/blob/25bc4d7a21752587dd6f964ab49a0a2722acbcba/model/inp/ww3_grid.inp#L808)

A good test to see if the boundary conditions are working as they should is to deactivate the wind forcing in your child grid simulation, and let the run go with the input wave boundary conditions only, then you should see the wave heights propagating into your domain, slowly 'filling' it.

**Reply by EmilioEchevarria** [reply]
I just re-read your message where you mentioned you only want boundaries in the south, east and north edges. In that case modify your ww3_grid.inp file like this:
```
110 20.5 1 0 -41
151 20.5 0 1 34
151 54.5 -1 0 41
```

**Reply by abayhaqi** [reply]
I modified ww3_grid.inp file of my regional wave as the referred example. However, when I run it, I got the error message.  Is there something else that I should modify?

[ww3_grid.txt](https://github.com/NOAA-EMC/WW3/files/12434420/ww3_grid.txt)
![run_ww3grid](https://github.com/NOAA-EMC/WW3/assets/49505651/67f9865c-191a-4ae1-8be9-734b5c3fb74e)

**Reply by EmilioEchevarria** [reply]
Please comment out (with the $ symbol) lines 63, 126, 193 and 605 from your ww3_grid.inp file and try again

**Reply by abayhaqi** [reply]
Thank you, It works. 
Then I tried to do what you suggested, I deactivated wind forcing in ww3_shel.inp . 
But, the  message comes with ERROR in W3IOBC. Does it means that the nested file can not be recognized?

**Reply by EmilioEchevarria** [reply]
It could be, W3IOBC is the subroutine that reads (or writes) boundary conditions files. Could you please paste the full error message?

**Reply by abayhaqi** [reply]
Here my full error message as follows:
 
WAVEWATCH III calculating for 2023/07/01 00:00:00 UTC at 10:27:21

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT   1   7 NOT ACTIVE SEA POINT (0)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT   1   3 NOT ACTIVE SEA POINT (0)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT   5   3 NOT ACTIVE SEA POINT (1)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT   9   3 NOT ACTIVE SEA POINT (1)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT  13   3 NOT ACTIVE SEA POINT (1)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT  17   3 NOT ACTIVE SEA POINT (1)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT  21   3 NOT ACTIVE SEA POINT (1)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT  25   3 NOT ACTIVE SEA POINT (1)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT  29   3 NOT ACTIVE SEA POINT (1)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT  33   3 NOT ACTIVE SEA POINT (1)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT  37   3 NOT ACTIVE SEA POINT (1)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT  41   3 NOT ACTIVE SEA POINT (1)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT  45   3 NOT ACTIVE SEA POINT (1)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT  49   3 NOT ACTIVE SEA POINT (1)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT  53   3 NOT ACTIVE SEA POINT (1)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT  57   3 NOT ACTIVE SEA POINT (1)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT  61   3 NOT ACTIVE SEA POINT (1)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT  65   3 NOT ACTIVE SEA POINT (1)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT  69   3 NOT ACTIVE SEA POINT (1)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT  73   3 NOT ACTIVE SEA POINT (1)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT  77   3 NOT ACTIVE SEA POINT (1)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT  81   3 NOT ACTIVE SEA POINT (1)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT  85   3 NOT ACTIVE SEA POINT (1)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT  89   3 NOT ACTIVE SEA POINT (1)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT  93   3 NOT ACTIVE SEA POINT (1)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT  97   3 NOT ACTIVE SEA POINT (1)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT 101   3 NOT ACTIVE SEA POINT (1)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT 105   3 NOT ACTIVE SEA POINT (1)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT 109   3 NOT ACTIVE SEA POINT (1)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT 113   3 NOT ACTIVE SEA POINT (1)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT 117   3 NOT ACTIVE SEA POINT (1)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT 121   3 NOT ACTIVE SEA POINT (1)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT 125   3 NOT ACTIVE SEA POINT (1)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT 129   3 NOT ACTIVE SEA POINT (1)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT 133   3 NOT ACTIVE SEA POINT (1)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT 137   3 NOT ACTIVE SEA POINT (1)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT 141   3 NOT ACTIVE SEA POINT (1)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT 145   3 NOT ACTIVE SEA POINT (1)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT 149   3 NOT ACTIVE SEA POINT (1)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT 153   3 NOT ACTIVE SEA POINT (1)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT 157   3 NOT ACTIVE SEA POINT (1)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT 161   3 NOT ACTIVE SEA POINT (0)

 *** WAVEWATCH III ERROR IN W3IOBC :
     POINT  43  0.151000E+03  0.205000E+02 NOT LOCATED IN GRID

EXTCDE MPI_ABORT, IEXIT=    12

--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 0 in communicator MPI_COMM_WORLD
with errorcode 12.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
--------------------------------------------------------------------------

**Reply by EmilioEchevarria** [reply]
I see. And can you please share your mask file? (`mapsta.inp`)

**Reply by abayhaqi** [reply]
Hi @EmilioEchevarria ,
Actually, I dont feel I have that file. I just try to run with the default setting and it turned OK when I run it before.

**Reply by abayhaqi** [reply]
I have this mask information file which is generated from Gridgen 

[s0_korean_maskinfo.txt](https://github.com/NOAA-EMC/WW3/files/12437075/s0_korean_maskinfo.txt)

**Reply by EmilioEchevarria** [reply]
Hi @abayhaqi, well no, you should try to adapt whatever configuration you find to your specific case. The mask you have at the moment looks like this:
<img src="https://github.com/NOAA-EMC/WW3/assets/36142422/39ca2982-4bb3-4d5c-bb27-9225cb2a19c0" width="40%">
So it has 1s for sea points and 0s for land points, you're missing the 2s for active boundary points (see [here](https://github.com/NOAA-EMC/WW3/blob/25bc4d7a21752587dd6f964ab49a0a2722acbcba/model/inp/ww3_grid.inp#L804-L809), read the description)

You need to edit your mask file to have 2s in the places where you want boundary conditions:
<img src="https://github.com/NOAA-EMC/WW3/assets/36142422/715ab36a-6aa4-46c3-9254-8881a80e193a" width="40%">

[s0_korean_maskinfo_BND.txt](https://github.com/NOAA-EMC/WW3/files/12449894/s0_korean_maskinfo_BND.txt)

**Reply by abayhaqi** [reply]
Hi @EmilioEchevarria 
Thank you for your explanation. I also have tried to make 2s in mask file. But, I still get an error when I run the ww3_grid. 
I posted my problem here [https://github.com/NOAA-EMC/WW3/discussions/1059](see).
Is there anything else should be revised?
Thank you

**Reply by abayhaqi** [reply]
This is the link for my problem just in case the previous one doesn't work 
[https://github.com/NOAA-EMC/WW3/discussions/1059#discussion-5565450](url)

**Reply by EmilioEchevarria** [reply]
It means there's something wrong with your grid input file (ww3_grid.inp), the format is not right. It seems to happen when WW3 tries to read Output boundary points... maybe try un-commenting line 654 in your ww3_grid.inp file?

**Reply by abayhaqi** [reply]
I have tried it, but it's still error. Is it because of line 621-624? (please see attachment). I am wondering whether I put it in the correct way
![ww3_grid](https://github.com/NOAA-EMC/WW3/assets/49505651/7b0828e3-ad1e-4849-881b-bf3181dfde92)

**Reply by EmilioEchevarria** [reply]
It could be. What are you trying to achieve with that exactly? Try commenting them out

**Reply by abayhaqi** [reply]
Yeah, It works. Now my ww3_grid.inp reads the mask file correctly. 
then I tried to run the model with deactivating wind forcing in ww3_shel.inp. However, I got this message.
Do you mind giving me insight on what this error means?
![error_message1](https://github.com/NOAA-EMC/WW3/assets/49505651/7325e866-636a-48e4-9c4f-f26dd01d9c44)

**Reply by EmilioEchevarria** [reply]
W3IORS deals with IO of restart file(s), there's something wrong there. Please share your ww3_shel.inp file

**Reply by abayhaqi** [reply]
Here I attach my wwe_shel.inp file
[ww3_shel_.txt](https://github.com/NOAA-EMC/WW3/files/12451108/ww3_shel_.txt)

**Reply by EmilioEchevarria** [reply]
By the way, NSPEC=800 is your number of frequencies * your number of directions (32*25), NSEA should be the number of 'sea' points in your grid, which I would expect it to be around 12k given the mask file you sent, I don't know why WW3 is expecting 7350 in your case. Please review your grid definitions, files, sizes, etc

**Reply by abayhaqi** [reply]
Actually, I run the global model, which produced nest.ww3, with 1 degree resolution, while the regional resolution is 0.25. Does that matter?

**Reply by EmilioEchevarria** [reply]
That should be fine. And you did run `ww3_grid` for your child grid and that's the `mod_def.ww3` that you're using, right?

**Reply by abayhaqi** [reply]
I used different ww3_grid for global and regional model. Or should it be the same? because I thought, there are different resolution and setting, I should make ww3_grid for regional again.

**Reply by EmilioEchevarria** [reply]
Run `ww3_grid` for the regional grid again and double check your input grids (it seems you're only using bathy and mask for now) and how you define them in ww3_grid.inp. And yes, global and regional should be different, independent runs

**Reply by abayhaqi** [reply]
Hi @EmilioEchevarria 
Just want to say thank you, your responses helped me a lot.
Now my simulation is working. 
Just one question, I am just wondering, What does point 3 (3: Point excluded from grid) really mean when we modify the mask file?
Thank you

**Reply by EmilioEchevarria** [reply]
Hi @abayhaqi, no problem at all, I'm glad you managed to get it working :)

A "3" in your mask file means that that point will be excluded from the calculations, and therefore ignored. It's not the same as a land point though. 

Feel free to close any issues you opened if they have been resolved. 
Good luck,
Emilio

**Reply by abayhaqi** [reply]
Hi @EmilioEchevarria , 
By the way, I am also wondering, why my global wave height simulation results have extremely high values in the north pole.
I used "RECT" "T" "SMPL" in grid.inp with full domain from -89.5 to 89.5
Do you mind to give me insight how to treat the north pole for global simulation?

**Reply by mickaelaccensi** [reply]
you have to reduce your cfl time step because the distances in km for each grid cell are going smaller when are closer to the pole for a same resolution in degree. Is your resolution at 0.5deg ?

**Reply by abayhaqi** [reply]
Hi @mickaelaccensi ,
My resolution is 1 degree

**Reply by mickaelaccensi** [reply]
so your DX at the northern latitude (lat=89.5) for a spatial resolution at 1deg is : 
dx=1*cos(89.5*3.14/180)*1852*60
dx=1057m

To estimate your CFL timestep Tcfl, the formula is : (I've considered that your lower frequency is 0.0373Hz)
Tcfl = dx / (g / (Fmin*4*Pi) ) with the constants Pi=3,14 and g=9.8m/s²;
Tcfl = 1057 / (9.8 / (0.0373*4*3.14) )
Tcfl = 50s

so your dtcfl should be set a bit lower, like :
dtcfl=45s
and dtmax, 3 times dtcfl, so :
dtmax=135s

As you can see, having a global grid going really close to the pole will require a very model timestep, that's why we usually reduce the global grid around latitudes 80N and then we create a curvilinear grid for the North pole, which has regular dx and dy in km units.

**Reply by abayhaqi** [reply]
Hi @mickaelaccensi 
Thank you so much for your explanation. It helped me.
Now my result is fine with reasonable values.

**Reply by mickaelaccensi** [reply]
ok great ! so I close the ticket

**Comment by nYifeiLee** [comment]
Dear @EmilioEchevarria ，
I went through a series of preparations for WW3 self-nesting based on your conversation with @abayhaqi .Now my large nest.ww3 file has been generated. But while ww3_shel in a small area, an error occurred in the program.I used python code to change the "1" to "2" of the boundary in the mask file of the small area.But the error was still reported.Can you help me see where my problem is appearing?Below is my input file and mask file.
Large scale grid:
[ww3_grid.txt](https://github.com/user-attachments/files/15632599/ww3_grid.txt)
[ww3_shel.txt](https://github.com/user-attachments/files/15632621/ww3_shel.txt)
[rect.mask.txt](https://github.com/user-attachments/files/15632660/rect.mask.txt)
Sarge scale grid:
[ww3_grid.txt](https://github.com/user-attachments/files/15633263/ww3_grid.txt)
[ww3_shel.txt](https://github.com/user-attachments/files/15633288/ww3_shel.txt)
[rect1.mask.txt](https://github.com/user-attachments/files/15633309/rect1.mask.txt)
ERROR:
![Ubuntu 22 - VMware Workstation 2024_6_6 22_02_46](https://github.com/NOAA-EMC/WW3/assets/167869457/f67e1e35-90ae-4322-be2a-9e9979a0cd27)

**Comment by walidchikhi** [comment]
@JMSnyder1980 , 
I am currently working for the first time on running WW3 over Algeria, using one grid covering the entire Mediterranean and another with higher resolution focused on the Algerian coast. However, I am facing some issues with the nesting. Do you happen to have any  test cases that I could use to reproduce the work over Algeria ? 
Warm regards ,
